### DISCUSSION TITLE
Annotate type with model validator

### DISCUSSION THREAD

**Discussion by ImogenBits** [pydantic/pydantic#6779] [Feature Request]
I'm not sure if this a feature request or just something that I can do already but haven't figured out how to. My basic problem is that I have a bunch of models that essentially look like this:
```py
class MyModel(BaseModel):
    some_list: list[int]
    some_index: int
```
With the idea being that `some_index` indexes into `some_list`. I can easily annotate that `some_index` is at least 0, but I haven't been able to figure out how to make pydantic validate that it's less than the list length.

To me it seems like the way to do this would be a model validator since you need access to both fields. But this becomes pretty annoying if you have multiple models with various index structures like this:
```py
class Base(BaseModel):
    some_list: list[int]

class BareIndex(Base):
    index: int

class IndexList(Base):
    index_list: list[int]

class IndexDict(Base):
    index_dict: dict[str, int]

...
```
Each of these would need their own model validator.

What I'd love to have would be something like this:
```py
class Base(BaseModel):
    some_list: list[int]

def check_index(self, val: int):
    assert isinstance(self, BaseModel)
    if not (0 <= val < len(some_list)):
        raise ValueError("Index is not a valid index into `some_list`")

Index = Annotated[int, ModelValueValidator(check_index, mode="after")]

class BareIndex(Base):
    index: Index

class IndexList(Base):
    index_list: list[Index]

class IndexDict(Base):
    index_dict: dict[str, Index]

...
```

I.e. this new model validator type would not only get the model it's validating but also a particular value of the type with the idea being that it's not really validating the model as such but just using it to further validate some value in it. These would then obviously all be called after other validation has been complete to make sure that whatever fields it needs for the validation are already populated.

Is there some way to do this already? The best I've been able to come up with is defining a model validator in the base class and manually recursing the type annotations to find `Index` fields, but that seems very error prone and like it'd duplicate a lot of the work that pydantic is doing already? If not and this is something that looks reasonable I'd be happy to help work on an implementation of this.

**Comment by tahayk** [comment]
Hey @ImogenBits 

not sure to fully understand your requirements, but I think you are looking for Before&After validators defined here: https://docs.pydantic.dev/2.5/concepts/validators/

Another way to solve this is by making 1 model and use Optional:

```python
class Base(BaseModel):
    some_list: list[int]
    index: Optional[int]
    index_list: Optional[list[int]]
    index_dict: Optional[dict[str, int]]

    @model_validator(mode="after")
    def check(cls, values):
        assert values.index is not None or values.index_list is not None or values.index_dict is not None,\
                   "at least 1 valid index should be provided"
        val = values.index if values.index is not None \
                 else values.index_list[0] if values.index_list is not None and len(values.index_list)>0 \
                 else values. index_dict["key"]
        assert 0<=val<=len(some_list), "index must be less or equal to the list's length"
        return values


```

**Comment by odravison** [comment]
What I understand from @ImogenBits was the same what I'm looking for, using an annotated type (IDK if it is the name for it) `MyCustomType = Annotated[int, someFunctionThatCheckMyInt]` and then using it in every class' attribute that I want to.

**Comment by odravison** [comment]
I found what I was looking for [here](https://docs.pydantic.dev/latest/concepts/validators/)
